### PR TITLE
Push types to type stack when calling a native

### PR DESCRIPTION
**Pull request by dariorussi**
## Description 

Make paranoid mode for native calls work correctly.
We may or may not enable paranoid mode but the change seems safe

## Test Plan 
Added test

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by amnn** [comment]
@dariorussi, to put me out of my misery, can you point me to where the equivalent operation is done for non-native entry points?

**Comment by dariorussi** [comment]
> @dariorussi, to put me out of my misery, can you point me to where the equivalent operation is done for non-native entry points?

lol - you won't find it because if you think about it the stack is empty on regular calls, the problem here is simply with natives where we push args on the stack (few lines above the changes) but we do not the relative types. Besides the fact that paranoid checks are sprinkled all over so it's hard to find a simple story to follow

**Comment by amnn** [comment]
So I saw/understood how things are set up if the interpreter is calling back into itself, but I guess the bit I don't understand is why the operand (and therefore also type) stack(s) are empty at the start of a regular call -- how do we pass the parameters to the entry function? (Or do we not check that actual params match the signature for the entry function?)

**Comment by tnowacki** [comment]
> So I saw/understood how things are set up if the interpreter is calling back into itself, but I guess the bit I don't understand is why the operand (and therefore also type) stack(s) are empty at the start of a regular call -- how do we pass the parameters to the entry function? (Or do we not check that actual params match the signature for the entry function?)

IIRC: For the bytecode operation `Call`, we have all the args on the stack. But, for a normal, non-native call, we eventually bind those args to local variables. And so for entry points (which has always been a function call) we bypass the bit of every putting those values on the stack and we take the args passed in and put them right into local variables. 
But! For natives, we don't have this local variable mechanism and we just expect the native function to pop them off the stack. So when natives are used as an entry point, we have this separate mechanism to take the args passed in and push them on the stack before calling the native function.

In short, the flow for a normal call and natives is a bit different at the point of entry.
We could unify them if we really wanted by making both just mimic the `Call` instruction but...

**Review by tnowacki — APPROVED** [review]
